8299684: (bf) JNI direct buffer functions with large capacity behave unexpectedly

bplb opened the pull request:
Remove cast in `JNI::NewDirectByteBuffer`of `long` capacity to `int`, modify the constructor in question to accept a `long` capacity, and verify in the constructor that the  capacity does not overflow `int` range, throwing IAE If it does.

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change requires CSR request [JDK-8299759](https://bugs.openjdk.org/browse/JDK-8299759) to be approved

### Issues
 * [JDK-8299684](https://bugs.openjdk.org/browse/JDK-8299684): (bf) JNI direct buffer functions with large capacity behave unexpectedly
 * [JDK-8299759](https://bugs.openjdk.org/browse/JDK-8299759): (bf) JNI direct buffer functions with large capacity behave unexpectedly (**CSR**)


### Reviewers
 * [David Holmes](https://openjdk.org/census#dholmes) (@dholmes-ora - **Reviewer**)
 * [Alan Bateman](https://openjdk.org/census#alanb) (@AlanBateman - **Reviewer**)


### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/11873/head:pull/11873` \
`$ git checkout pull/11873`

Update a local copy of the PR: \
`$ git checkout pull/11873` \
`$ git pull https://git.openjdk.org/jdk pull/11873/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 11873`

View PR using the GUI difftool: \
`$ git pr show -t 11873`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/11873.diff">https://git.openjdk.org/jdk/pull/11873.diff</a>

</details>

PaulSandoz commented on src/java.base/share/classes/java/nio/Direct-X-Buffer.java.template:
```diff
@@ -239,6 +240,17 @@ class Direct$Type$Buffer$RW$$BO$
+        try {
+            return Math.toIntExact(capacity);
+        } catch (ArithmeticException ignore) {
+            throw new IllegalArgumentException("capacity " + capacity
```
I think we should mention the source, that the capacity value originates from a native call to NewDirectByteBuffer, since it that will not appear in any frames of the repoted stack trace.

bplb replied:
> I think we should mention the source, that the capacity value originates from a native call to NewDirectByteBuffer [...]

So you mean in the IAE message itself?

PaulSandoz replied:
Yes, enhance the message with more detail

PaulSandoz reviewed (commented):
Requires an update to the JNI spec and therefore a CSR?

bplb commented:
> Requires an update to the JNI spec and therefore a CSR?

Yes, I think both.

dholmes-ora commented:
I think it will look strange to update the JNI spec to say that although the function takes a jlong as the requested capacity, it will throw IAE if the capacity is larger than a jint! The mismatch between the JNI and Java code made me dig into the history here to see what JSR-51 had to say about this, but unfortunately it didn't say much. What I did find in

https://bugs.openjdk.org/browse/JDK-4496703

was a very definitive comment about only supporting 32-bit (ie int) based direct buffers, but with a glimmer of hope 

> "A future revision of the specification is likely to address this problem". 

So perhaps the JNI side was prepared for this unrealized future? But even so it should have specified what happens if the jlong capacity exceeds the value of a jint

I thought perhaps a special OutOfMemoryError (akin to the "Requested array size exceeds VM limit" OOME) could be thrown - and that would not require a JNI spec change - but it is a stretch. Though also note that if this requires a JNI spec change then it cannot be backported without jumping through some serious JCP Maintenance Release hoops.

AlanBateman commented:
I checked ancient history and the issue goes back to JDK 1.4 (when JSR-51 defined this API). It's surprising that it hasn't been reported/noticed before now.

Yes, this will need an update to the JNI spec to say that it throws if capacity is negative or greater than Integer.MAX_VALUE.  It will look a bit strange in the spec when the parameter is jlong but we can't change it. As regards what to throw then the current proposal to throw IAE seems okay as the function already throws it for negative values that are >= Integer.MIN_VALUE, e.g.

```
Exception in thread "main" java.lang.IllegalArgumentException: capacity < 0: (-1 < 0)
	at java.base/java.nio.Buffer.createCapacityException(Buffer.java:282)
	at java.base/java.nio.Buffer.<init>(Buffer.java:245)
	at java.base/java.nio.ByteBuffer.<init>(ByteBuffer.java:298)
	at java.base/java.nio.ByteBuffer.<init>(ByteBuffer.java:306)
	at java.base/java.nio.MappedByteBuffer.<init>(MappedByteBuffer.java:113)
	at java.base/java.nio.DirectByteBuffer.<init>(DirectByteBuffer.java:177)
	at Test.newDirectByteBuffer(Native Method)
	at Test.main(Test.java:7)
```

The compatibility impact of the change should be minimal, at least I can't imagine anyone depending on the current broken behavior.

For the change, clampCapacity only needs to be called once, not twice. The exception message is a bit strange. If cap is < Integer.MIN_VALUE then it would be better to have it consistent with the existing message above. Also "in upcall from JNI NewDirectByteBuffer" should be better if it didn't include "in upcall".

We will need to add a test for this and decide whether to place it. It seems there aren't any tests for these JNI functions in hotspot//jtreg/runtime/jni, maybe we start a new locations for tests in test/jdk/java/nio/jni.

AlanBateman commented:
/csr

dholmes-ora commented on make/test/JtregNativeJdk.gmk:
```diff
@@ -78,9 +78,11 @@ ifeq ($(call isTargetOs, windows), true)
   BUILD_JDK_JTREG_LIBRARIES_CFLAGS_libAsyncInvokers := -I$(TEST_LIB_NATIVE_SRC)
 
   BUILD_JDK_JTREG_LIBRARIES_LIBS_libTracePinnedThreads := jvm.lib
+  BUILD_JDK_JTREG_LIBRARIES_LIBS_libNewDirectByteBuffer := -ljava -lc
```
This is not correct for Windows

bplb replied:
Do you have a suggestion?

dholmes-ora replied:
```
BUILD_JDK_JTREG_LIBRARIES_LIBS_libNewDirectByteBuffer := jvm.lib
```
as per the preceding line.

dholmes-ora replied:
Sorry, my bad, that should be:
```
BUILD_JDK_JTREG_LIBRARIES_LIBS_libNewDirectByteBuffer := $(WIN_LIB_JAVA)
```
the Windows equivalent of `-ljava`.

bplb replied:
Thanks. So changed in 82d9e6f1f6782b180943d1136881d58dd2852e58.

dholmes-ora commented on make/test/JtregNativeJdk.gmk:
```diff
@@ -78,9 +78,11 @@ ifeq ($(call isTargetOs, windows), true)
 else
   BUILD_JDK_JTREG_LIBRARIES_LIBS_libstringPlatformChars := -ljava
   BUILD_JDK_JTREG_LIBRARIES_LIBS_libDirectIO := -ljava
+  BUILD_JDK_JTREG_LIBRARIES_LIBS_libNewDirectByteBuffer := -ljava -lc
```
Why do we need `-lc`?

bplb replied:
No. Removed in d2386a4b73f0fa4eb9cc278b3cbb1578e6cb4b55.

dholmes-ora commented on src/java.base/share/classes/java/nio/Direct-X-Buffer.java.template:
```diff
@@ -239,6 +240,19 @@ class Direct$Type$Buffer$RW$$BO$
 
+    // Constrain the capacity to int range throwing IAE if not possible
+    //
+    private static int clampCapacity(long capacity) {
```
I think `checkCapacity` would be better - clamp suggests the value gets clamped rather than throwing an exception.

bplb replied:
So changed in d2386a4b73f0fa4eb9cc278b3cbb1578e6cb4b55.

dholmes-ora commented on src/java.base/share/classes/java/nio/Direct-X-Buffer.java.template:
```diff
@@ -239,6 +240,19 @@ class Direct$Type$Buffer$RW$$BO$
+    // Constrain the capacity to int range throwing IAE if not possible
+    //
+    private static int clampCapacity(long capacity) {
+        try {
+            return Math.toIntExact(capacity);
```
Why not the more direct and obvious:
```
if (capacity < 0) {
  throw ...
} else if (capacity > Integer.MAX_VALUE) {
  throw ...
}
```

bplb replied:
So changed in d2386a4b73f0fa4eb9cc278b3cbb1578e6cb4b55.

dholmes-ora commented on src/java.base/share/classes/java/nio/Direct-X-Buffer.java.template:
```diff
@@ -239,6 +240,19 @@ class Direct$Type$Buffer$RW$$BO$
+        } catch (ArithmeticException ignore) {
+            String msg = "JNI NewDirectByteBuffer passed illegal capacity: "
+                + capacity + (capacity < Integer.MIN_VALUE
+                ? " < Integer.MIN_VALUE" : " > Integer.MAX_VALUE");
```
Why the check against `MIN_VALUE`?

bplb replied:
It was the lower bound for `toIntExact()` but not relevant any more.

dholmes-ora commented on test/jdk/java/nio/jni/NewDirectByteBuffer.java:
```diff
@@ -0,0 +1,80 @@
+    private static final long[] LEGAL_CAPACITIES = {
+        0L,
+        1L,
+        (long)Integer.MAX_VALUE - 1,
+        (long)Integer.MAX_VALUE
```
Won't such large capacities trigger OOME?

AlanBateman replied:
> Won't such large capacities trigger OOME?

The malloc might fail, in which case the value of NULL will be passed to NewDirectByteBuffer. It needs to check the return from malloc.

bplb replied:
Fixed in d2386a4b73f0fa4eb9cc278b3cbb1578e6cb4b55.

AlanBateman commented:
I agree with David that it would be clearer and simpler to check if capacity is negative or greater than Integer.MAX_VALUE. There other helper methods called from this code and they are named checkXXX so it would be good to keep that consistent if you can.

We don't have other tests for this JNI function in the test tree so having the new tests be a complete unit test would be good. This means checking the position, limit, and capacity of the ByteBuffer and testing that it's a direct buffer. If you make it a JUnit test then you'll be able to paramatrized test with value sources, and use the assertions API when checking each of the buffer properties.

For the native code used by the test then it would be clearer if allocBigBuffer were renamed to newDirectByteBuffer. Also I think you'll need to check the value from malloc in case it fails. You'll need another native function to free the memory and have the tests invoke the free in a finally block so that memory doesn't accumulate as the test cycles through the different buffer sizes. Right now, I assume the tests don't run on 32-bit as it mallocs two 2Gb areas. Once the test is re-worded then you can get a sense as to how much memory it needs and whether you need a @requires in the test description to ensue that the test system has sufficient memory.

bplb commented:
> If you make it a JUnit test then you'll be able to paramatrized test with value sources, and use the assertions API when checking each of the buffer properties.

All points addressed in d2386a4b73f0fa4eb9cc278b3cbb1578e6cb4b55 except converting to a JUnit test.

dholmes-ora commented on test/jdk/java/nio/jni/NewDirectByteBuffer.java:
```diff
@@ -75,6 +101,7 @@ public static void main(String[] args) {
     // See libNewDirectByteBuffer.c for implementations.
-    private static native ByteBuffer allocBigBuffer(long size);
-    private static native long getLongCapacity(ByteBuffer buf);
+    private static native ByteBuffer newDirectByteBuffer(long size);
+    private static native long getDirectBufferCapacity(ByteBuffer buf);
+    private static native void freeDirectBufferMemory(ByteBuffer buf);
```
Naming nit: ByteBuffer versus Buffer

bplb replied:
Fixed in 95fd6652148e1b740af599081b3614c54656c8b4 (overlooked in previous commit).

dholmes-ora commented on test/jdk/java/nio/jni/NewDirectByteBuffer.java:
```diff
@@ -42,31 +50,49 @@ public class NewDirectByteBuffer {
 
     public static void main(String[] args) {
         System.out.println("--- Legal Capacities ---");
         for (long cap : LEGAL_CAPACITIES) {
```
I would still expect this to throw OOME for the very large values, so the OOME needs to be caught so that the test continues.

bplb replied:
Fixed in 82d9e6f1f6782b180943d1136881d58dd2852e58.

AlanBateman reviewed (commented):
The update looks quite good and maybe some day this can be converted from a psvm test to a junit test.

bplb commented on src/java.base/share/classes/java/nio/Direct-X-Buffer.java.template:
```diff
@@ -181,12 +181,28 @@ class Direct$Type$Buffer$RW$$BO$
+        if (capacity < 0) {
+            throw new IllegalArgumentException
+                ("JNI NewDirectByteBuffer passed capacity < 0: ("
+                + capacity + " < 0)");
```
I was matching `Buffer::createCapacityException` but I agree with you. Remove in 82d9e6f1f6782b180943d1136881d58dd2852e58.

bplb commented on src/java.base/share/classes/java/nio/Direct-X-Buffer.java.template:
```diff
@@ -181,12 +181,28 @@ class Direct$Type$Buffer$RW$$BO$
+        } else if (capacity > Integer.MAX_VALUE) {
+            throw new IllegalArgumentException
+                ("JNI NewDirectByteBuffer passed capacity > Integer.MAX_VALUE: ("
+                + capacity + " > Integer.MAX_VALUE)");
```
Removed in 82d9e6f1f6782b180943d1136881d58dd2852e58.

bplb commented on test/jdk/java/nio/jni/NewDirectByteBuffer.java:
```diff
@@ -0,0 +1,107 @@
+        -1L,
+        (long)Integer.MAX_VALUE + 1L,
+        3_000_000_000L,
+        5_000_000_000L
```
Added in 82d9e6f1f6782b180943d1136881d58dd2852e58.

bplb commented on test/jdk/java/nio/jni/libNewDirectByteBuffer.c:
```diff
@@ -0,0 +1,65 @@
+    // if throwing the OOME fails
+    void* addr = malloc(size);
+    if (addr == NULL) {
+        jclass rtExCls = (*env)->FindClass(env, "java/lang/OutOfMemoryError");
```
Thanks for catching that. The name `rtExCls` is as you interpreted and is residual from an unpublished version. Fixed in 82d9e6f1f6782b180943d1136881d58dd2852e58.

bplb commented:
> The update looks quite good and maybe some day this can be converted from a psvm test to a junit test.

`JUnit` or `TestNG`?

LanceAndersen commented:
> > The update looks quite good and maybe some day this can be converted from a psvm test to a junit test.
> 
> `JUnit` or `TestNG`?

It sounds like the suggested way forward is to use junit for new tests and for extra credit convert existing testng tests when/if they need updated

bplb commented on test/jdk/java/nio/jni/NewDirectByteBuffer.java:
```diff
@@ -0,0 +1,107 @@
+ * @bug 8299684
+ * @summary Verify that JNI NewDirectByteBuffer throws IllegalArgumentException
+ * if the capacity is negative or greater than Integer::MAX_VALUE
+ * @run main/native NewDirectByteBuffer
```
Added **@requires** in 7f99faeebaea1dc55b592146aa3ba612bdb63d9e.

bplb commented:
> If you make it a JUnit test then you'll be able to paramatrized test with value sources, and use the assertions API when checking each of the buffer properties.

Converted to a JUnit test in 7f99faeebaea1dc55b592146aa3ba612bdb63d9e.
 
> Once the test is re-worded then you can get a sense as to how much memory it needs and whether you need a @requires in the test description to ensue that the test system has sufficient memory.

Added **@requires** constraining to 64-bit arch and memory at least 8GB in 7f99faeebaea1dc55b592146aa3ba612bdb63d9e.

AlanBateman commented on test/jdk/java/nio/jni/NewDirectByteBuffer.java:
```diff
@@ -0,0 +1,102 @@
+        Assertions.assertEquals(0L, buf.position(),
+            "Buffer::position returned unexpected value");
+        Assertions.assertEquals(capacity, buf.limit(),
+            "Buffer::limit returned unexpected value");
```
For completeness, I think checkBuffer should also set/get some elements.

AlanBateman reviewed (commented):
The test looks much better now but I think can be improved a bit more.

bplb commented on test/jdk/java/nio/jni/NewDirectByteBuffer.java:
```diff
@@ -0,0 +1,102 @@
+
+import org.junit.jupiter.api.Assertions;
+import org.junit.jupiter.params.ParameterizedTest;
+import org.junit.jupiter.params.provider.ValueSource;
```
Fixed in 2fe819ba1fd929eaaa6cc31caee38e217006c3f0.

bplb commented on test/jdk/java/nio/jni/NewDirectByteBuffer.java:
```diff
@@ -0,0 +1,102 @@
+ * @summary Verify that JNI NewDirectByteBuffer throws
+ * IllegalArgumentException if the capacity is negative or greater than
+ * Integer::MAX_VALUE
+ * @requires (sun.arch.data.model == "64" & os.maxMemory >= 8g)
```
So changed in 2fe819ba1fd929eaaa6cc31caee38e217006c3f0.

bplb commented on test/jdk/java/nio/jni/NewDirectByteBuffer.java:
```diff
@@ -0,0 +1,102 @@
+
+    @ParameterizedTest
+    @ValueSource(longs = {Long.MIN_VALUE, (long)Integer.MIN_VALUE - 1L, -1L,
+        (long)Integer.MAX_VALUE + 1L, 3_000_000_000L, 5_000_000_000L,
```
They were in the initial bug report.

bplb commented on test/jdk/java/nio/jni/NewDirectByteBuffer.java:
```diff
@@ -0,0 +1,102 @@
+                checkBuffer(buf, capacity);
+            }
+        } catch (OutOfMemoryError ignored) {
+            // Ignore the error
```
So changed in 2fe819ba1fd929eaaa6cc31caee38e217006c3f0.

bplb commented on test/jdk/java/nio/jni/NewDirectByteBuffer.java:
```diff
@@ -0,0 +1,102 @@
+                    freeDirectByteBufferMemory(buf);
+                }
+            });
+        } catch (OutOfMemoryError ignored) {
```
Apparently it is not needed so removed in 38ea7997be2be59e2cea97c7dabf929210c0edfa.

dholmes-ora replied:
No wonder I couldn't find this change! The catch of OOME was added because I still believe that trying to allocate a buffer of capacity Integer.MAX_VALUE will throw OOME (at least some of the time). If that happens then the test should not fail.

AlanBateman replied:
No objection to having the test pass if malloc fails but it would be better if the native newDirectByteBuffer were changed to just call JNI NewDirectByteBuffer with a given address so that the malloc + JNI function aren't in the same native method. That allows the test to distinguish malloc returning NULl from NewDirectByteBuffer throwing OOME. The native method freeDirectByteBufferMemory can be split too so that there is native method get the direct buffer address. With these changes, the test can use Unsafe allocateMemory/freeMemory (changing to FFM APIs in the future) or have native methods that do the malloc/free. The more basic methods means the test can be improved,  e.g. checkBuffer doesn't currently test that GetDirectBufferAddress returns the address that the DBB was created to wrap.

AlanBateman reviewed (approved):
I think this looks okay. If you are doing any further edits then the try-finally OOME can be removed from the illegalCapacities test.

AlanBateman commented on test/jdk/java/nio/jni/NewDirectByteBuffer.java:
```diff
@@ -0,0 +1,136 @@
+    @ValueSource(longs = {0L, 1L, (long)Integer.MAX_VALUE/2,
+        (long)Integer.MAX_VALUE - 1, (long)Integer.MAX_VALUE})
+    void legalCapacities(long capacity) {
+        long addr = UNSAFE.allocateMemory(capacity);
```
I think David was asking for the test to continue when malloc fails. Now that the test is changed to use Unsafe.allocateMemory it means it has to catch OOME.

bplb replied:
Added catch of OOME in f9350695c3a6f90e3acdf2176d78c06406e49387.

AlanBateman commented on test/jdk/java/nio/jni/libNewDirectByteBuffer.c:
```diff
@@ -0,0 +1,50 @@
+Java_NewDirectByteBuffer_getDirectBufferAddress
+    (JNIEnv *env, jclass cls, jobject buf)
+{
+    return (jlong)(*env)->GetDirectBufferAddress(env, buf);
```
The cast to jlong should be okay for 64-bit. For 32-bit then it will likely need a double cast, as in ((jlong)(int)( .. )) but maybe okay for now as you've restricted the test to 64-bit.

bplb replied:
Leaving it as-is for now.

dholmes-ora replied:
FTR this was the wrong choice as the test is still built on 32-bit (and fails to compile).

dholmes-ora commented on test/jdk/java/nio/jni/NewDirectByteBuffer.java:
```diff
@@ -0,0 +1,143 @@
+        Long.MAX_VALUE})
+    void illegalCapacities(long capacity) {
+        assertThrows(IllegalArgumentException.class, () -> {
+            long addr = UNSAFE.allocateMemory(capacity);
```
How is `allocateMemory` going to respond to these illegal capacity requests?

If we expect `newDirectByteBuffer` to throw then can't we use a dummy buffer value as it will never be used?

AlanBateman replied:
> How is `allocateMemory` going to respond to these illegal capacity requests?
> 
> If we expect `newDirectByteBuffer` to throw then can't we use a dummy buffer value as it will never be used?

For illegalCapacities it should call allocateMemory with some fixed value, say 1, not capacity. Surprised the allocateMemory isn't failing for the test inputs.

bplb replied:
Changed allocation size to `1` in `illegalCapacities` in 3b1eef25b53dbeb36e4b3a8d6c279a5e0f5a5ae3.

dholmes-ora reviewed (approved):
Nothing further from me. Thanks

bplb commented:
/integrate

iklam commented:
This PR is causing causing x86 build failures in GitHub actions. I filed https://bugs.openjdk.org/browse/JDK-8300942

bplb commented:
> This PR is causing causing x86 build failures in GitHub actions. I filed https://bugs.openjdk.org/browse/JDK-8300942

PR #12157 has been created to address this problem.
